File: server/app/api/schemas/exports.py
from datetime import datetime
from typing import Annotated, Literal
from uuid import UUID

from pydantic import AwareDatetime, BaseModel, ConfigDict, Field

from app.api.schemas.fares import ResponseMeta


class ExportJobCreateRequest(BaseModel):
    model_config = ConfigDict(populate_by_name=True)

    subscription_id: UUID = Field(alias="subscriptionId")
    format: Literal["csv", "json"]
    range_start: AwareDatetime = Field(alias="rangeStart")
    range_end: AwareDatetime = Field(alias="rangeEnd")
    idempotency_key: Annotated[
        str,
        Field(
            alias="idempotencyKey",
            min_length=8,
            max_length=80,
            pattern=r"^[A-Za-z0-9._:-]+$",
        ),
    ]


class ExportJobPublic(BaseModel):
    model_config = ConfigDict(populate_by_name=True)

    id: UUID
    subscription_id: UUID | None = Field(alias="subscriptionId")
    format: Literal["csv", "json"]
    scope: Literal["canonical_query"] = "canonical_query"
    status: Literal["pending", "running", "succeeded", "failed", "expired", "deleting"]
    range_start: datetime = Field(alias="rangeStart")
    range_end: datetime = Field(alias="rangeEnd")
    snapshot_at: datetime = Field(alias="snapshotAt")
    attempt: int
    max_attempts: int = Field(alias="maxAttempts")
    processed_rows: int = Field(alias="processedRows")
    row_count: int | None = Field(alias="rowCount")
    size_bytes: int | None = Field(alias="sizeBytes")
    checksum_sha256: str | None = Field(alias="checksumSha256")
    file_name: str | None = Field(alias="fileName")
    error_code: str | None = Field(alias="errorCode")
    error_message: str | None = Field(alias="errorMessage")
    created_at: datetime = Field(alias="createdAt")
    started_at: datetime | None = Field(alias="startedAt")
    completed_at: datetime | None = Field(alias="completedAt")
    expires_at: datetime | None = Field(alias="expiresAt")
    download_ready: bool = Field(alias="downloadReady")


class ExportJobListResponse(BaseModel):
    model_config = ConfigDict(populate_by_name=True)

    meta: ResponseMeta
    items: list[ExportJobPublic]
    has_more: bool = Field(alias="hasMore")
    next_cursor: str | None = Field(alias="nextCursor")

File: web/src/api/exports.ts
import { apiRequest } from '@/api/client'

export type ExportFormat = 'csv' | 'json'
export type ExportStatus = 'pending' | 'running' | 'succeeded' | 'failed' | 'expired' | 'deleting'

export interface PriceExportJob {
  id: string
  subscriptionId: string | null
  format: ExportFormat
  scope: 'canonical_query'
  status: ExportStatus
  rangeStart: string
  rangeEnd: string
  snapshotAt: string
  attempt: number
  maxAttempts: number
  processedRows: number
  rowCount: number | null
  sizeBytes: number | null
  checksumSha256: string | null
  fileName: string | null
  errorCode: string | null
  errorMessage: string | null
  createdAt: string
  startedAt: string | null
  completedAt: string | null
  expiresAt: string | null
  downloadReady: boolean
}

export interface PriceExportList {
  meta: { mode: 'live' | 'demo'; generatedAt: string }
  items: PriceExportJob[]
  hasMore: boolean
  nextCursor: string | null
}

export interface CreatePriceExportInput {
  subscriptionId: string
  format: ExportFormat
  rangeStart: string
  rangeEnd: string
  idempotencyKey: string
}

export const exportQueryKeys = {
  all: ['price-exports'] as const,
  list: (subscriptionId: string) => [...exportQueryKeys.all, 'list', subscriptionId] as const,
}

export async function getPriceExports(subscriptionId: string, cursor?: string): Promise<PriceExportList> {
  const query = new URLSearchParams({ subscriptionId, limit: '50' })
  if (cursor) query.set('cursor', cursor)
  return apiRequest<PriceExportList>(`/exports?${query.toString()}`)
}

export async function createPriceExport(input: CreatePriceExportInput): Promise<PriceExportJob> {
  return apiRequest<PriceExportJob>('/exports', {
    method: 'POST',
    body: JSON.stringify(input),
  })
}

export async function deletePriceExport(jobId: string): Promise<void> {
  await apiRequest(`/exports/${jobId}`, { method: 'DELETE' })
}

export function priceExportDownloadUrl(jobId: string): string {
  return `/api/exports/${encodeURIComponent(jobId)}/download`
}
